semcore/dropdown/src/AbstractDropdown.tsx

```diff
@@ -0,0 +1,347 @@
+import { localizedMessages } from './translations/__intergalactic-dynamic-locales';
+import { isFocusInside, setFocus } from '@semcore/utils/lib/use/useFocusLock';
+import focusSourceEnhance from '@semcore/utils/lib/enhances/focusSourceEnhance';
+import { DropdownProps } from './index';
+
+type AbstractDDProps = {
+  visible: boolean;
+  highlightedIndex: number;
+  selectedIndex: number;
+  placement: DropdownProps['placement'];
+  inlineActions: boolean;
+  disablePortal: boolean;
+  ignorePortalsStacking: boolean;
+  interaction: DropdownProps['interaction'];
+  timeout?: number | [number, number];
+};
+
+export const enhance = {
+  uid: uniqueIDEnhancement(),
+  getI18nText: i18nEnhance(localizedMessages),
+  focusSourceRef: focusSourceEnhance(),
+};
+
+type Enhances<T extends { [key: string]: (...args: any) => any }> = {
+  [K in keyof T]: ReturnType<T[K]>[K];
+};
+
+export const selectedIndexContext = React.createContext(0);
+
+export abstract class AbstractDropdown extends Component<
+  AbstractDDProps,
+  {},
+  {},
+  Enhances<typeof enhance>
+> {
+  protected abstract role: 'menu' | 'listbox';
+
+  popperRef = React.createRef<HTMLElement>();
+  triggerRef = React.createRef<HTMLElement>();
+  menuRef = React.createRef<HTMLElement>();
+
+  itemProps: any[] = [];
+  itemRefs: HTMLElement[] = [];
+
+  highlightedItemRef = React.createRef<HTMLElement>();
+
+  prevHighlightedIndex = 0;
+
+  uncontrolledProps() {
+    return {
+      selectedIndex: null,
+      highlightedIndex: [
+        null,
+        (index: number) => {
+          this.handlers.selectedIndex(index);
+        },
+      ],
+      visible: null,
+    };
+  }
+
+  get childRole() {
+    if (this.role === 'listbox') {
+      return 'option';
+    }
+
+    return 'menuitem';
+  }
+
+  handleClickTrigger = (e: React.SyntheticEvent) => {
+    const { interaction } = this.asProps;
+
+    if (interaction === 'none') return false;
+
+    e.preventDefault();
+    this.handlers.visible(true);
+
+    if (this.role === 'menu') {
+      setTimeout(() => {
+        const { highlightedIndex } = this.asProps;
+        const element = this.itemRefs[highlightedIndex];
+        element?.focus();
+      }, 0);
+    }
+  };
+
+  getTriggerProps() {
+    const { size, uid, disablePortal, visible, getI18nText } = this.asProps;
+
+    return {
+      size,
+      id: `igc-${uid}-trigger`,
+      focusHint: visible && !disablePortal ? getI18nText('triggerHint') : undefined,
+      'aria-haspopup': 'true',
+      'aria-expanded': visible ? 'true' : 'false',
+      onClick: this.handleClickTrigger,
+      ref: this.triggerRef,
+    };
+  }
+
+  getListProps() {
+    const { size, uid } = this.asProps;
+    const triggerId = this.triggerRef.current?.id;
+
+    return {
+      size,
+      index: this.asProps.highlightedIndex,
+      tabIndex: -1,
+      ref: this.menuRef,
+      id: `igc-${uid}-list`,
+      role: this.role,
+      'aria-labelledby': triggerId,
```

Comment:
listbox should have the same label as the Select.Trigger (ideally)
but if that's too complicated, we can make it a static text string (like before)

Reply:
I made ideally option, but check it please

Reply:
works like magic!!!